photon-client/src/main/scala/com/linkedin/photon/ml/data/avro/NameAndTermFeatureMapUtils.scala

```diff
@@ -0,0 +1,153 @@
+      val featureSetPath = new Path(outputDir, featureSectionKey)
+      saveNameAndTermsAsTextFiles(featureSet, featureSetPath, sc)
+    }
+  }
+
+  /**
+   * Write a [[NameAndTerm]] feature [[RDD]] to HDFS as a single text file.
+   *
+   * @param nameAndTermRDD The [[NameAndTerm]] feature [[RDD]] to be written
+   * @param outputPath The HDFS path to which write the [[NameAndTerm]] feature [[RDD]]
+   * @param sc The Spark context
+   */
+  private def saveNameAndTermsAsTextFiles(
+    nameAndTermRDD: RDD[NameAndTerm],
+      outputPath: Path,
+      sc: SparkContext): Unit = {
+
+    val hadoopFS = FileSystem.get(sc.hadoopConfiguration)
+    val tmpPath = new Path(outputPath.getParent, s"${outputPath.getName}-tmp")
+
+    // saveAsTextFile in a temporary folder
+    hadoopFS.delete(tmpPath, true)
+    // save result uncompressed for human-readable output
+    sc.hadoopConfiguration.set("mapred.output.compress", "false")
+    nameAndTermRDD.map(_.toString).saveAsTextFile(tmpPath.toString)
+
+    // Merge files in temporary folder into final output
+    hadoopFS.delete(outputPath, true)
+    FileUtil.copyMerge(
+      hadoopFS, tmpPath,
+      hadoopFS, outputPath,
+      true, sc.hadoopConfiguration, null
+    )
+  }
+
+  /**
+   * Parse the [[Map]] of feature section key to [[NameAndTerm]] feature [[RDD]] from text files on HDFS.
+   *
+   * @param inputDir The input HDFS directory
+   * @param featureSectionKeys The set of feature section keys to look for from the input directory
+   * @param sc The Spark context
+   * @return A [[Map]] of feature section key to [[NameAndTerm]] feature [[RDD]] parsed from text files on HDFS
+   */
+  protected[ml] def readNameAndTermFeatureMapFromTextFiles(
+    inputDir: Path,
+    featureSectionKeys: Set[String],
+    sc: SparkContext): Map[String, RDD[NameAndTerm]] = {
+
+    featureSectionKeys
+      .map { featureSectionKey =>
+        val inputPath = new Path(inputDir, featureSectionKey)
+        val nameAndTermFeatureSet = readNameAndTermRDDFromTextFiles(inputPath, sc)
+        (featureSectionKey, nameAndTermFeatureSet)
+      }
+      .toMap
+  }
+
+  /**
+   * Read a [[RDD]] of [[NameAndTerm]] from the text files within the input path.
+   *
+   * @param inputPath The input path
+   * @param sc The Spark context
+   * @return The [[RDD]] of [[NameAndTerm]] read from the text files of the given input path
+   */
+  private def readNameAndTermRDDFromTextFiles(inputPath: Path, sc: SparkContext): RDD[NameAndTerm] = {
+
+    sc.textFile(inputPath.toString)
+      .map { string =>
+        string.split("\t") match {
+          case Array(name, term) => NameAndTerm(name, term)
+          case Array(name) => NameAndTerm(name, "")
+          case other => throw new UnsupportedOperationException(
+            s"Unexpected entry $string when parsing it to NameAndTerm, " +
+              s"after splitting by tab the expected number of tokens is 1 or 2, but found ${other.length}}.")
```

Comment:
Error message shouldn't include "after splitting by tab" - implementation details like that shouldn't be exposed to users in error messages.